WebApp/src/websocket.ts

```diff
@@ -0,0 +1,146 @@
+                // from: from connection id
+                // to: to connection id
+                // data: any message data structure
+
+                const msg = JSON.parse(event.data);
+                if (!msg || !this) {
+                    return;
+                }
+
+                console.log(msg);
+
+                switch (msg.type) {
+                    case "connect":
+                        this.onConnect(ws);
+                        break;
+                    case "disconnect":
+                        this.onDisConnect(ws, msg.data);
+                        break;
+                    case "offer":
+                        this.onOffer(ws, msg.data);
+                        break;
+                    case "answer":
+                        this.onAnswer(ws, msg.data);
+                        break;
+                    case "cadicate":
+                        this.onCadicate(ws, msg.data);
+                        break;
+                    default:
+                        break;
+                }
+            };
+        });
+    }
+
+    private onConnect(ws: WebSocket){
+        const connectionId: string = uuid();
+        const connectionIds = getOrCreateConnectionIds(ws);
+        connectionIds.add(connectionId);
+        ws.send(JSON.stringify({connectionId:connectionId}));
+    }
+
+    private onDisConnect(ws: WebSocket, message: any){
+        const connectionIds = clients.get(ws);
+        const connectionId = message.connectionId as string;
+        connectionIds.delete(connectionId);
+        connectionPair.delete(connectionId);
+    }
+
+    private onOffer(ws: WebSocket, message: any){
+        const connectionId = message.connectionId as string;
+        const newOffer = new Offer(message.sdp, Date.now());
+        offers.set(connectionId, newOffer);
+        connectionPair.set(connectionId, [ws, null]);
+        clients.forEach((_v, k) => k.send(JSON.stringify({from:connectionId, to:"", type:"offer", data:newOffer})));
+    }
+
+    private onAnswer(ws: WebSocket, message: any){
+        const connectionId = message.connectionId as string;
+        const connectionIds = getOrCreateConnectionIds(ws);
+        connectionIds.add(connectionId);
+        const newAnswer = new Answer(message.sdp, Date.now());
+        answers.set(connectionId, newAnswer);
+
+        const pair = connectionPair.get(connectionId);
+        const otherSessionId = pair[0];
+        connectionPair.set(connectionId, [otherSessionId, ws]);
+
+        const mapCandidates = candidates.get(otherSessionId);
+        if (mapCandidates) {
+          const arrayCandidates = mapCandidates.get(connectionId);
+          for (const candidate of arrayCandidates) {
+            candidate.datetime = Date.now();
+          }
+        }
+        clients.forEach((_v, k) => k.send(JSON.stringify({from:connectionId, to:"", type:"answer", data:newAnswer})));
+    }
+
+    private onCadicate(ws: WebSocket, message: any){
```

Comment:
Like what @karasusan said, this should be `onCandidate` too

Reply:
I fixed it